src/madness/world/worldrmi.cc

```diff
@@ -332,10 +335,19 @@ namespace madness {
     void RMI::RmiTask::huge_msg_handler(void *buf, size_t /*nbytein*/) {
         const size_t* info = (size_t *)(buf);
         int nword = HEADER_LEN/sizeof(size_t);
-        int src = info[nword];
-        size_t nbyte = info[nword+1];
-
-        RMI::task_ptr->hugeq.push_back(std::make_pair(src,nbyte));
+        const int src = info[nword];
+        const size_t nbyte = info[nword+1];
+        const int tag = info[nword+2];
+
+        // extra dose of paranoia: assert that we never process so many huge messages
+        // that the tag wraparound somewhere becomes possible ...
+        // the worst case is where only one node sends huge messages to every node in the communicator
+        // AND it has enough threads to use up all tags
+        // NB list::size() is O(1) in c++11, but O(N) in older libstdc++
+        MADNESS_ASSERT(ThreadPool::size() < RMI::RmiTask::unique_tag_period() ||
+                       RMI::task_ptr->hugeq.size() <
+                           RMI::RmiTask::unique_tag_period() / comm.Get_size());
```

Comment:
I did not Realize the cloned communicator was just for rmi. We will still
need to take another copy for the rest of the madness runtime.

On Sep 11, 2016 7:41 PM, "Eduard Valeyev" <notifications@github.com> wrote:

> In src/madness/world/worldrmi.cc
> <https://github.com/m-a-d-n-e-s-s/madness/pull/209#discussion_r78305070>:
>
> > -        int src = info[nword];
> > -        size_t nbyte = info[nword+1];
> > -
> > -        RMI::task_ptr->hugeq.push_back(std::make_pair(src,nbyte));
> > +        const int src = info[nword];
> > +        const size_t nbyte = info[nword+1];
> > +        const int tag = info[nword+2];
> > +
> > +        // extra dose of paranoia: assert that we never process so many huge messages
> > +        // that the tag wraparound somewhere becomes possible ...
> > +        // the worst case is where only one node sends huge messages to every node in the communicator
> > +        // AND it has enough threads to use up all tags
> > +        // NB list::size() is O(1) in c++11, but O(N) in older libstdc++
> > +        MADNESS_ASSERT(ThreadPool::size() < RMI::RmiTask::unique_tag_period() ||
> > +                       RMI::task_ptr->hugeq.size() <
> > +                           RMI::RmiTask::unique_tag_period() / comm.Get_size());
>
> 772684d
> <https://github.com/m-a-d-n-e-s-s/madness/commit/772684d43b73df3d9e832fc259f660edb4a2da06>
> introduced an RMI-specific communicator, and thus removes the tag collision
> issue (but I did introduce a separate range to keep thing kosher even
> without the change).
>
> the check is correct: with nonpreemptive scheduler and nothing truly
> exotic happening on the programming side there is no way for 1 thread to
> work on 2 huge messages. So if # of threads < tag period the RMI tag
> wraparound cannot occur. If that is not the case we have to be very
> pessimistic (remember that sender has no way of checking how many tags on
> that node a live) ... think one node's threads firing roughly same number
> of huge messages to each other node
>
> again: the test is correct in my opinion.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/m-a-d-n-e-s-s/madness/pull/209/files/8b27ed90e83e475b462d26f0155c8c0156a49ea3#r78305070>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHMgPY1BKX70LOA5G3kVkhtnFeTFlsDEks5qpJGLgaJpZM4J1Poj>
> .
>